meeting-notes/2024-12-11-Books-DB-and-Books-CRUD-contracts.md

```diff
@@ -0,0 +1,77 @@
+# 2024-12-11 Books DB & Books CRUD contracts
+
+## Валидация
+1. Книга не может быть без автора - проверить, что авторов в добавляемой книге > 1
+2. Определить максимальную длину строк в полях книги, автора и ограничить
+3.  Валидация должна быть на уровне базы, на уровне бэка и на уровне фронта
+
+
+## Что проверить/посмотреть?
+1. Как хранить язык книги - enum в виде строки (если это пофиксили)
+2. Доклад про B-Tree индекс
+3. Как сейчас реализовано удаление книг? Если удалить книгу, удалится ли автор? Если удалить автора, удалится ли книга?
+4. Как возвращаем сообщение об ошибке? (RFC?)
+5. Краткие коды языков для контрактов по книжкам ("eng" скорее всего может быть просто "e")
+
+
+## Улучшения, изменения
+1. Используем только такие HTTP методы: GET, POST, DELETE. Не используем PUT
+2. Меняем url по примеру:
+/api/books - POST создает
+/api/books/1/edit - POST
+/api/books/1/soft-delete - DELETE
+/api/books/1/hard-delete - DELETE
+/api/books - GET - все
+/api/books/1 - GET - одна книга
+/api/books/1/authors - GET
+3. При создании книги возвращаем { "newBookId": 1 }
+4. При удалении возвращаем { “isDeleted”: true }, потому что Cypress для тестов на фронте нужно что-то возвращать для эндпоинта удаления
+5. При многократном удалении одного и того же результата не возвращаем ошибку, а только { “isDeleted”: true } - идемпотентность
+6. Написать E2E на Karate для всех CRUD операций
+7. Прописать в контрактах все респонсы
+8. Реализовать редактирование для всех полей, в т.ч. автор, язык
+9. Добавить примеры значений полей в базе данных (чтобы было понятно, какие данные лежат в реальной используемой таблице - отразить в диаграмме в виде дополнительной таблицы (или если можно - в ней же, если есть такой функционал))
+10. Добавить все поля, которых не хватает в доке, но они уже есть в реализации: tenant_id, deleted_at_utc и т.д.
+
+## Каких контрактов не хватает?
+1. Запрос на одну книгу (по id)
+
```

Comment:
```suggestion
2. Поддержка тенантов согласно https://learn.microsoft.com/en-us/ef/core/miscellaneous/multitenancy, для кейса хранения колонки `TenantId` в каждой таблице.
```

Reply:
are you sure it should be in this section?